### PR TITLE
Add XLA Profiler section to docs

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Part of #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7

### CODE REVIEW

File: pytorch_lightning/profiler/xla.py [hunk 1]
```diff
@@ -12,32 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """XLA Profiler will help you debug and optimize training workload performance for your models using Cloud TPU
```
Comment (rohitgr7) [inline pytorch_lightning/profiler/xla.py:14]: maybe move this inside the class definition for consistency with other components.

File: docs/source/advanced/profiler.rst [hunk 2]
```diff
@@ -204,3 +204,36 @@ Use::
 Or::
 
     python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'
+
+
+XLA Profiling
+-------------
+
+`XLA Profiler <https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.profiler.XLAProfiler.html#pytorch_lightning.profiler.XLAProfiler>`_ will help you debug and optimize training
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:212]:
```suggestion
:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
```

File: docs/source/advanced/profiler.rst [hunk 3]
```diff
@@ -204,3 +204,36 @@ Use::
...
+##############################
+Manual capture via TensorBoard
+##############################
+
+The following instructions are for capturing trace from a running program:
+
+0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
+help you with the Cloud TPU setup with the required installations
+
+1. Start a TensorBoard Server. You could view the TensorBoard output at http://localhost:9001 on your local machine, and then open the
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:224]:
```suggestion
1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
```

File: docs/source/advanced/profiler.rst [hunk 4]
```diff
@@ -204,3 +204,36 @@ Use::
...
+
+XLA Profiling
+-------------
+
+`XLA Profiler <https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.profiler.XLAProfiler.html#pytorch_lightning.profiler.XLAProfiler>`_ will help you debug and optimize training
+workload performance for your models using Cloud TPU performance tools.
+
+##############################
+Manual capture via TensorBoard
+##############################
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:217]:
```suggestion
Manual capture via TensorBoard
^^^^^^^^^^^^^^^^^^^^^^^^
```

File: pytorch_lightning/profiler/xla.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/profiler/xla.py]: Updated!

File: docs/source/advanced/profiler.rst [hunk 5]
```diff
@@ -204,3 +204,35 @@ Use::
 Or::
 
     python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'
+
+
+XLA Profiling
+-------------
+
+:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
+workload performance for your models using Cloud TPU performance tools.
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:213]:
```suggestion
workload performance for your models using Cloud TPU performance tools.

.. code-block:: python
    # by passing a string
    trainer = Trainer(..., profiler="xla")

    # or by passing an instance
    from pytorch_lightning.profiler.xla import XLAProfiler

    profiler = XLAProfiler(port=?)
    trainer = Trainer(..., profiler=profiler)
```

what is port in `XLAProfiler`?

File: docs/source/advanced/profiler.rst [hunk 6]
```diff
@@ -204,3 +204,35 @@ Use::
...
+help you with the Cloud TPU setup with the required installations
+
+1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
+``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile
+
+.. code-block:: bash
+
+    tensorboard --logdir ./tensorboard --port 9001
+
+2. Once the code you'd like to profile is running, click on ``CAPTURE PROFILE`` button. You could enter
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:230]:
```suggestion
The following instructions are for capturing traces from a running program:

0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
help you with the Cloud TPU setup with the required installations

1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile

.. code-block:: bash

    tensorboard --logdir ./tensorboard --port 9001

2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. You could enter
```
Grammarly 🙈

File: docs/source/advanced/profiler.rst (hunk 5 above)
Comment (kaushikb11) [reply docs/source/advanced/profiler.rst:213]: Updated! Added it to docstring

File: docs/source/advanced/profiler.rst [hunk 7]
```diff
@@ -212,6 +212,18 @@ XLA Profiling
 :class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
 workload performance for your models using Cloud TPU performance tools.
 
+.. code-block:: python
+
+    # by passing the `XLAProfiler` alias
+    trainer = Trainer(..., profiler="xla")
+
+    # or by passing an instance
+    from pytorch_lightning.profiler.xla import XLAProfiler
```
Comment (ananthsub) [inline docs/source/advanced/profiler.rst:221]: should we make the profiler imports all be exposed at the lightning.profiler module? this would be consistent with what we do for other components. fyi @daniellepintz

Comment (kaushikb11) [reply docs/source/advanced/profiler.rst]: Yup, updated! Thanks

File: docs/source/advanced/profiler.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaushikb11) [reply docs/source/advanced/profiler.rst:213]: We mean it in the generic manner here. The profiling would work the same for all running stage processes. As it's focus is mainly towards performance summary and finding bottlenecks in the XLA operations.

File: pytorch_lightning/profiler/xla.py [hunk 8]
```diff
@@ -63,8 +36,13 @@ class XLAProfiler(BaseProfiler):
...
 
     def __init__(self, port: int = 9012) -> None:
-        """This Profiler will help you debug and optimize training workload performance for your models using Cloud
-        TPU performance tools."""
+        """XLA Profiler will help you debug and optimize training workload performance for your models using Cloud
+        TPU performance tools.
+
+        Args:
+            port : the port to start the profiler server on. An exception is
+                raised if the provided port is invalid or busy. Default: ``9012``.
```
Comment (rohitgr7) [inline pytorch_lightning/profiler/xla.py:44]:
```suggestion
            port : the port to start the profiler server on. An exception is
                raised if the provided port is invalid or busy.
```
defaults are added by sphinx

File: docs/source/advanced/profiler.rst [hunk 9]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
+``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile
+
+.. code-block:: bash
+
+    tensorboard --logdir ./tensorboard --port 9001
+
+2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. You could enter
+``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, you could enter
```
Comment (rohitgr7) [inline docs/source/advanced/profiler.rst:243]: quick ques: are these ports used differently?

File: docs/source/advanced/profiler.rst [hunk 10]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+    # or by passing an instance
+    from pytorch_lightning.profiler import XLAProfiler
+
+    profiler = XLAProfiler(port=9012)
+    trainer = Trainer(..., profiler=profiler)
+
+
+Manual capture via TensorBoard
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:228]: @rohitgr7 started with standardizing the titles, let's make it capitalized :)) sorry for the nit ^^

File: docs/source/advanced/profiler.rst [hunk 11]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+The following instructions are for capturing traces from a running program:
+
+0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
+help you with the Cloud TPU setup with the required installations
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:233]:
```suggestion
help you with the Cloud TPU setup with the required installations.
```

File: docs/source/advanced/profiler.rst [hunk 12]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
+``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:236]: the link could escaped/formatted

File: docs/source/advanced/profiler.rst [hunk 13]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+.. code-block:: bash
+
+    tensorboard --logdir ./tensorboard --port 9001
+
+2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. You could enter
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:242]:
```suggestion
2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. Enter
```

File: docs/source/advanced/profiler.rst (hunk 9 above)
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:243]:
```suggestion
``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, enter
```

File: docs/source/advanced/profiler.rst [hunk 14]
```diff
@@ -204,3 +204,47 @@ Use::
...
+``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, you could enter
+the number of milliseconds for the profiling duration, and click ``CAPTURE``
+
+3. Make sure the code is running, while you are trying to capture the traces. Also, it would lead to better
+performance insights if the profiling duration is longer than the step time
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:247]:
```suggestion
3. Make sure the code is running while you are trying to capture the traces. Also, it would lead to better
performance insights if the profiling duration is longer than the step time.
```

File: docs/source/advanced/profiler.rst [hunk 15]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+4. Once the capture is finished, the page will refresh and you could browse through the insights using the
+``Tools`` dropdown at the top left
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:250]:
```suggestion
4. Once the capture is finished, the page will refresh and you can browse through the insights using the
``Tools`` dropdown at the top left
```

File: pytorch_lightning/profiler/xla.py [hunk 16]
```diff
@@ -63,8 +36,13 @@ class XLAProfiler(BaseProfiler):
     }
 
     def __init__(self, port: int = 9012) -> None:
-        """This Profiler will help you debug and optimize training workload performance for your models using Cloud
-        TPU performance tools."""
+        """XLA Profiler will help you debug and optimize training workload performance for your models using Cloud
+        TPU performance tools.
+
+        Args:
+            port : the port to start the profiler server on. An exception is
```
Comment (awaelchli) [inline pytorch_lightning/profiler/xla.py:43]:
```suggestion
            port: the port to start the profiler server on. An exception is
```

**Review by awaelchli — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source/advanced/profiler.rst [hunk 17]
```diff
@@ -204,3 +204,47 @@ Use::
...
+
+Manual capture via TensorBoard
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following instructions are for capturing traces from a running program:
+
+0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
+help you with the Cloud TPU setup with the required installations.
+
+1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:235]: Can we add a gif to show the steps?

File: docs/source/advanced/profiler.rst [hunk 18]
```diff
@@ -204,3 +204,47 @@ Use::
...
+``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile
+
+.. code-block:: bash
+
+    tensorboard --logdir ./tensorboard --port 9001
+
+2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. Enter
+``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, enter
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:243]:
```suggestion
``localhost:9001`` (default port for XLA Profiler) as the Profile Service URL. Then, enter
```

Comment (tchaton) [reply docs/source/advanced/profiler.rst:243]: not sure ?

File: docs/source/advanced/profiler.rst (hunk 9 above)
Comment (akihironitta) [reply docs/source/advanced/profiler.rst]: As I read the Google Cloud blog post, it should be the same, i.e., 9001. https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm

File: docs/source/advanced/profiler.rst (hunk 18 above)
Comment (akihironitta) [reply docs/source/advanced/profiler.rst:251]: https://github.com/PyTorchLightning/pytorch-lightning/pull/11436#discussion_r783869996

File: docs/source/advanced/profiler.rst [hunk 19]
```diff
@@ -214,6 +214,50 @@ To visualize the profiled operation, you can either:
...
+
+.. code-block:: python
+
+    # by passing the `XLAProfiler` alias
+    trainer = Trainer(..., profiler="xla")
+
+    # or by passing an instance
+    from pytorch_lightning.profiler import XLAProfiler
+
+    profiler = XLAProfiler(port=9001)
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:231]: @kaushikb11 I updated the port from 9012 to 9001 here in 37b1601. Correct? If not, please revert the commit...

Comment (kaushikb11) [reply docs/source/advanced/profiler.rst:231]: It's right!
